File: Dataset/Leetcode/train/70/312.cpp
class Solution {
public:
    // 背包容量是n, 有无数种1阶和2阶台阶(商品), 求装满背包的排列总数
    int XXX(int n) {
        vector<int> dp(n + 1, 0);
        dp[0] = 1;
        // 求排列总数, 先遍历背包
        for(int j = 0; j <= n; ++j)
            for(int i = 1; i <= 2; ++i)
                if(j >= i)
                    dp[j] += dp[j - i];
        return dp.back();
    }
};

File: Dataset/Leetcode/test/3/312.java
 class Solution {
    public int XXX(String s) {
        HashMap str = new HashMap();
        int len = 0;
        int max = 0;
        int prevIndex = 0;
        int index = 0;
        for(int i = 0;i<s.length();i++){
            if(str.containsKey(s.charAt(i))){
                index = (int)str.get(s.charAt(i));
                if( index < prevIndex){
                    str.remove(index);
                    str.put(s.charAt(i),i);
                    continue;
                }
                len = i - prevIndex;
                prevIndex = index + 1;
                if( max < len){
                    max = len;
                }
                str.remove(index);
            }
            str.put(s.charAt(i),i);
        }
        len = s.length() - prevIndex;
        if( max < len){
            max = len;
        }
        return max;
    }
}

